src/Shared/ExceptionHandling.cs

```diff
@@ -301,22 +301,26 @@ internal static void DumpExceptionToFile(Exception ex)
                 if (s_dumpFileName == null)
                 {
                     Guid guid = Guid.NewGuid();
-                    string tempPath = Path.GetTempPath();
+                    string dumpDirEnv = Environment.GetEnvironmentVariable("MSBUILDDEBUGPATH");
+                    string dumpDir = !string.IsNullOrEmpty(dumpDirEnv)
+                            ? dumpDirEnv
+                            : Path.GetTempPath();
 
                     // For some reason we get Watson buckets because GetTempPath gives us a folder here that doesn't exist.
                     // Either because %TMP% is misdefined, or because they deleted the temp folder during the build.
-                    if (!Directory.Exists(tempPath))
+                    if (!Directory.Exists(dumpDir))
                     {
                         // If this throws, no sense catching it, we can't log it now, and we're here
                         // because we're a child node with no console to log to, so die
-                        Directory.CreateDirectory(tempPath);
+                        Directory.CreateDirectory(dumpDir);
                     }
 
-                    s_dumpFileName = Path.Combine(tempPath, "MSBuild_" + guid.ToString() + ".failure.txt");
+                    var pid = Process.GetCurrentProcess().Id;
```

Comment:
Why did you introduce a new variable here?

Reply:
I wanted to put PID into the filename. Makes it just a little easier to correlate the log file to other parts of the logging output without having to open each file to find the PID.